### PR TITLE
Made weighting consistent with original GCN implementation

### PR DESCRIPTION
**Pull request by jackd**
This PR makes weighting of samples consistent with the [original implementation](https://github.com/tkipf/gcn/blob/master/gcn/metrics.py#L4). Note this also makes results on `pubmed` consistent with those published.

There are other gcn examples that could use the same fix that I am happy to apply this to, and potentially non-gcn examples (I haven't looked into other implementations yet). I'm happy to update this PR to include fixes to those, but I am unsure if maintainers would prefer duplicate fixes or a dataset method to load normalized weights, e.g.

```python
class Citation(Dataset):
  ...
    def load_mask(self, split='train', dtype=np.bool):
        if split == 'train':
            mask = self.mask_tr
        elif split == 'validation':
            mask = self.mask_va
        elif split == 'test':
            mask = self.mask_te
        else:
            raise ValueError(f'split must be one of "train", "validation", "test", got {split}')
        return mask.astype(dtype)
    
    def load_sample_weights(self, split='train', normalize=True, dtype=np.float32):
        weights = self.load_mask(split=split, dtype=dtype)
        if normalize:
            weights /= np.count_nonzero(weights)
        return weights
```

### CODE REVIEW

File: examples/node_prediction/citation_gcn.py [hunk 1]
```diff
@@ -52,13 +59,13 @@
 model = Model(inputs=[x_in, a_in], outputs=gc_2)
 optimizer = Adam(lr=learning_rate)
 model.compile(optimizer=optimizer,
-              loss='categorical_crossentropy',
+              loss=CategoricalCrossentropy(reduction='sum'),
```
Comment (jackd) [inline examples/node_prediction/citation_gcn.py:62]: 'categorical_crossentropy' uses `AUTO` reduction, which defaults to `SUM_OVER_BATCH_SIZE`. AFAIK there is no way to use a weighted mean reduction, hence we normalize the weights and use `sum`.

**Comment by danielegrattarola** [comment]
Hi, 

thanks for spotting this issue, very nice catch! 

The fix is general enough that I think it should be a feature of the `Dataset` class itself, not just for `Citation`. 
I think it could have this general structure: 

- `mask_tr`, `mask_va`, `mask_te` attributes populated by the `read()` method, which is user-defined. All of these should be binary masks, should default to `None` if not specified by the user, and should not be mandatory for a Dataset. 
- `get_mask(self, split='train', dtype=np.bool)` method that returns the masks;
- `get_sample_weights(self, split='train', dtype=np.float32)` that returns the masks re-weighted;
- also a `get_mask_index(self, split='train', dtype=np.int)` method that returns `np.argwhere(mask)` could be useful

Do you want to take a shot at the implementation? If not, I can work on this and use this PR as working branch for implementing the feature.

Also, I will check online to see which examples need fixing.

Cheers, 
Daniele

**Comment by jackd** [comment]
I only really see this applying to `single` datasets - most multi-graph datasets have graphs that each fall into one split entirely, right? If you wanted to roll it out generally I'd put the attributes into `Graph` and make `Loader`s return datasets with `(inputs, labels, sample_weights)` where applicable.

I'll be tied up with some other stuff today, so if it's easiest for you to make the changes you want feel free to go ahead and do whatever you want :)

**Comment by danielegrattarola** [comment]
I have opted for only fixing the examples, I'll think about a longer term solution in the future. 

Thanks for the PR!
Daniele
